### ISSUE TITLE
CMake issues (sqlite3, llvm, threads)

### ISSUE THREAD

**Issue by myfreeweb** [mull-project/mull#401]
While working on a FreeBSD port (package) for mull, I discovered a few issues with the CMake build scripts:

- The `LLVM_ROOT` variable shouldn't be named that. Modern CMake (we have 3.12 right now) ignores it and prints a large warning, because `*_ROOT` variables are reserved for found packages.
- SQLite3 is not searched for and just assumed to live on the default linker/include paths
- When the installed LLVM has a `libLLVM.so` shared library (which is the correct way to ship LLVM in an OS distribution), the separate `LLVMAsmParser` etc. libraries in `target_link_libraries` cause the infamous "LLVM ERROR: inconsistency in registered CommandLine options" to appear, which indicates multiple versions of LLVM in a process (I guess both the shared libLLVM and static small libraries get linked).
- The pthread library is not marked as a dependency of libmull.

Here's my patches (minus the `LLVM_ROOT` thing), but you might want to do things differently (e.g. `IMPORTED_TARGET` is a relatively new feature and might not be present on the oldest CMake version you want to support so you might want to find some `FindSQLite` script instead of direct pkg-config calls; also I duplicated sqlite3 find in a couple files) and find some way to autodetect `LLVMAsmParser+LLVMOrcJIT+…` vs `LLVM` (or just provide a config option for that).

```diff
--- lib/CMakeLists.txt.orig	2018-10-01 09:02:27 UTC
+++ lib/CMakeLists.txt
@@ -1,3 +1,6 @@
+find_package(PkgConfig)
+pkg_check_modules(SQLITE3 sqlite3 IMPORTED_TARGET)
+
 set(mull_sources
   ConfigParser.cpp
   Config.cpp
@@ -106,16 +109,14 @@ add_library(mull
 
 set_source_files_properties(${mull_header_dirs} PROPERTIES HEADER_FILE_ONLY ON)
 
+find_package(Threads REQUIRED)
+
 target_link_libraries(mull
-  LLVMAsmParser
-  LLVMOrcJIT
-  LLVMSupport
-  LLVMOption
-  LLVMX86CodeGen
-  LLVMX86AsmParser
+  LLVM
   libclang
   clangTooling
-  sqlite3
+  PkgConfig::SQLITE3
+  Threads::Threads
 )
 target_include_directories(mull PUBLIC ${MULL_INCLUDE_DIRS})
 
--- tools/reporter/CMakeLists.txt.orig	2018-10-06 15:11:37 UTC
+++ tools/reporter/CMakeLists.txt
@@ -1,3 +1,6 @@
+find_package(PkgConfig)
+pkg_check_modules(SQLITE3 sqlite3 IMPORTED_TARGET)
+
 add_executable(mull-reporter
   WeakTestsReporter.h
   WeakTestsReporter.cpp
@@ -10,7 +13,7 @@ set_target_properties(mull-reporter PROPERTIES
 )
 target_link_libraries(mull-reporter
   LLVMSupport
-  sqlite3
+  PkgConfig::SQLITE3
 )
 target_include_directories(mull-reporter PUBLIC
   ${MULL_INCLUDE_DIRS}
--- unittests/CMakeLists.txt.orig	2018-10-06 15:09:25 UTC
+++ unittests/CMakeLists.txt
@@ -37,6 +37,9 @@ set(mull_unittests_sources
   ConfigParserTestFixture.h
 )
 
+find_package(PkgConfig)
+pkg_check_modules(SQLITE3 sqlite3)
+
 add_executable(MullUnitTests ${mull_unittests_sources})
 target_link_libraries(MullUnitTests
   mull
@@ -46,6 +49,7 @@ target_include_directories(MullUnitTests PUBLIC
   ${CMAKE_CURRENT_SOURCE_DIR}
   ${MULL_SOURCE_DIR}/googletest/include
   ${MULL_INCLUDE_DIRS}
+  ${SQLITE3_INCLUDE_DIRS}
 )
 set_target_properties(MullUnitTests PROPERTIES
   COMPILE_FLAGS ${MULL_CXX_FLAGS}
```

**Comment by myfreeweb** [comment]
> Good catch, I'm not curious how it ever worked

Well, the pthread error appeared only after I replaced the `LLVM*` libs with just `LLVM`… somehow the small-LLVM targets were depending on it?

> does it actually work on FreeBSD?

The report I got on the helloworld example shows applied mutations, so I guess it does :)

**Comment by AlexDenisov** [comment]
> Well, the pthread error appeared only after I replaced the LLVM* libs with just LLVM… somehow the small-LLVM targets were depending on it?

Yes, seems like this is the case. We should link this library explicitly, though.

> The report I got on the helloworld example shows applied mutations, so I guess it does :)

This is awesome, thank you for the great news :-D

**Comment by AlexDenisov** [comment]
@myfreeweb all the mentioned issues are resolved by now, see #402 and #403 for more details.

Thank you very much for the report. Feel free to reopen this issue if something is still not working.

**Comment by myfreeweb** [comment]
Thanks! Looks like you missed one — `LLVM_ROOT` is still called `LLVM_ROOT`

**Comment by AlexDenisov** [comment]
@myfreeweb, yes, you are right, it will come soon with another fix.

**Comment by AlexDenisov** [comment]
The last issue is also resolved: #404.
